adapter/src/ethereum.rs

```diff
@@ -1,101 +1,486 @@
+
+        if channel_status != 1.into() {
+            return Err(AdapterError::Configuration(
+                "channel is not Active on the ethereum network".to_string(),
+            ));
+        }
+
         Ok(true)
     }
 
-    fn session_from_token(&self, _token: &str) -> AdapterResult<String> {
-        // @TODO
-        Ok("hello".to_string())
+    fn session_from_token(&mut self, token: &str) -> AdapterResult<Session> {
+        let token_id = token.to_owned()[..16].to_string();
+        let result = self.tokens_verified.get(&token_id);
+        if result.is_some() {
+            return Ok(result.unwrap().to_owned());
+        }
+
+        let verified = match ewt_verify(&token) {
+            Ok(v) => v,
+            Err(e) => return Err(AdapterError::EwtVerifyFailed(format!("{}", e))),
+        };
+
+        let whoami = self.whoami()?;
+        if whoami != verified.from {
+            return Err(AdapterError::Configuration(
+                "token payload.id !== whoami(): token was not intended for us".to_string(),
+            ));
+        }
+
+        let sess = match &verified.payload.identity {
+            Some(identity) => {
+                let (_eloop, transport) =
+                    web3::transports::Http::new(&self.config.ethereum_network).unwrap();
+                let web3 = web3::Web3::new(transport);
+
+                let contract_address =
+                    Address::from_slice(self.config.ethereum_core_address.as_bytes());
+
+                let contract = Contract::from_json(
+                    web3.eth(),
+                    contract_address,
+                    include_bytes!("../contract/Identity.json"),
+                )
+                .unwrap();
+
+                let contract_query =
+                    contract.query("privileges", verified.from, None, Options::default(), None);
+                let priviledge_level: U256 = contract_query.wait().unwrap();
+
+                if priviledge_level == 0.into() {
+                    return Err(AdapterError::Authorization(
+                        "insufficient privilege".to_string(),
+                    ));
+                }
+                Session {
+                    era: verified.payload.era,
+                    uid: identity.to_owned(),
+                }
+            }
+            None => Session {
+                era: verified.payload.era,
+                uid: verified.from,
+            },
+        };
+
+        self.tokens_verified.insert(token_id, sess.clone());
+        Ok(sess)
+    }
+
+    fn get_auth(&mut self, validator: &ValidatorDesc) -> AdapterResult<String> {
+        match (self.tokens_for_auth.get(&validator.id), &self.wallet) {
+            (Some(token), Some(_)) => Ok(token.to_owned()),
+            (None, Some(wallet)) => {
+                let payload = Payload {
+                    id: validator.id.clone(),
+                    era: usize::try_from(Utc::now().timestamp()).expect("failed to parse utc now"),
```

Comment:
this is an actual OK case for `expect`! cause there's no realistic way it can fail